### PR TITLE
Apply flakehell to netket

### PR DESCRIPTION
**Pull request by PhilipVinc**
This adds `flakehell` as a dev tool to netket and solves most issues revealed by it.
It's a rebranded version of `pyflakes` and `pylint` that makes it easy to disable specific errors and add others.

It uncovered 3 few bugs, like wrong assertions raised that don't exist, inexistant arguments... those are addressed.

There is still a bug that it has uncovered that is not solved : JsonLog in append mode will overwrite the file.
I will address this in a separate PR.

most Issues were solved automatically (removed extra imports a bit everywhere) a few formatting issues and then a bunch where addressed manually.

To run flakehell manually you just need to do `flakehell lint netket`.

### CODE REVIEW

**Comment by PhilipVinc** [comment]
There are still a few issues (~4) that flakehell raises that i havent addressed because i need to think about it.
However, we might eventually turn it on as a pre-commit hook.

**Comment by femtobit** [comment]
Very nice!

>  However, we might eventually turn it on as a pre-commit hook.

I am definitely in favor of this. Can we already do this while explicitly ignoring the few known issues you've mentioned?

**Comment by PhilipVinc** [comment]
> I am definitely in favor of this. Can we already do this while explicitly ignoring the few known issues you've mentioned?

Yes, but that would mean adding another file (called a baseline, which is the list of past issues that should be ignored).
Addressing the few remaining issues is relatively little work that I can do it relatively soon and makes everything easier...

Also, i'd rather turn on the mandatory linter in a follow up pr

**Comment by PhilipVinc** [comment]
By the way, this PR also bumps the version of black we use to run black. to a more recent version (so that it plays better together with the linter. They sometimes disagree on what should be done)

File: netket/utils/array.py [hunk 1]
```diff
@@ -32,14 +31,13 @@ class HashableArray:
...
 
-    def __post_init__(self):
-        object.__setattr__(self, "wrapped", self.wrapped.copy())
-        if isinstance(self.wrapped, np.ndarray):
-            self.wrapped.flags.writeable = False
-        object.__setattr__(self, "_HashableArray__hash", hash(self.wrapped.tobytes()))
+    def __pre_init__(self, wrapped):
+        wrapped = wrapped.copy()
+        wrapped.flags.writeable = False
+        return (wrapped,), {}
```
Comment (femtobit) [inline netket/utils/array.py:37]: This will fail if `wrapped` is not an `ndarray` but a jax `DeviceArray`. (Because that one does not provide the `writeable` flag.) The `isinstance(self.wrapped, np.ndarray)` check needs to be added back.

File: netket/vqs/mc_state.py [hunk 2]
```diff
@@ -727,10 +724,10 @@ def grad_expect_operator_kernel(
...
 
     has_aux = mutable is not False
-    if not has_aux:
-        out_axes = (0, 0)
-    else:
-        out_axes = (0, 0, 0)
+    # if not has_aux:
+    #    out_axes = (0, 0)
+    # else:
+    #    out_axes = (0, 0, 0)
```
Comment (femtobit) [inline netket/vqs/mc_state.py:730]: What is the purpose of these changes?

Comment (PhilipVinc) [reply netket/vqs/mc_state.py:730]: that is unused code.
Flake8 complains it's unused.
I want to keep it to remind myself to get back to it (it's needed for models that have `has_aux=True`).

File: netket/hilbert/_deprecations.py [hunk 3]
```diff
@@ -22,21 +22,22 @@ def graph_to_N_depwarn(N, graph):
...
+            ``graph=_your_graph``, as they no longer take it from the hilbert space.
+            """
         )
 
         if N == 1:
             return graph.n_nodes
         else:
             raise ValueError(
-                "Graph object can only take one argumnent among N and graph (deprecated)."
+                "Graph object can only take one argumnent among N and graph"
```
Comment (femtobit) [inline netket/hilbert/_deprecations.py:39]:
```suggestion
                "Graph object can only take one argument among N and graph"
```

File: netket/hilbert/_deprecations.py [hunk 4]
```diff
@@ -22,21 +22,22 @@ def graph_to_N_depwarn(N, graph):
...
 
-			You can update your code by passing `N=_your_graph.n_nodes`.
-			If you are also using Ising, Heisemberg, BoseHubbard or GraphOperator
-			hamiltonians you must now provide them with the extra argument
-			``graph=_your_graph``, as they no longer take it from the hilbert space.
-			"""
+            You can update your code by passing `N=_your_graph.n_nodes`.
+            If you are also using Ising, Heisemberg, BoseHubbard or GraphOperator
+            hamiltonians you must now provide them with the extra argument
+            ``graph=_your_graph``, as they no longer take it from the hilbert space.
```
Comment (femtobit) [inline netket/hilbert/_deprecations.py:31]:
```suggestion
            If you are also using `Ising`, `Heisenberg`, `BoseHubbard` or `GraphOperator`
            Hamiltonians you must now provide them with the extra argument
            ``graph=_your_graph``, as they no longer take it from the Hilbert space.
```

File: netket/hilbert/_deprecations.py [hunk 5]
```diff
@@ -22,21 +22,22 @@ def graph_to_N_depwarn(N, graph):
         warn_deprecation(
             r"""
             The ``graph`` argument for hilbert spaces has been deprecated in v3.0.
-			It has been replaced by the argument ``N`` accepting an integer, with 
-			the number of nodese in the graph. 
+            It has been replaced by the argument ``N`` accepting an integer, with
+            the number of nodese in the graph.
```
Comment (femtobit) [inline netket/hilbert/_deprecations.py:26]:
```suggestion
            It has been replaced by the argument ``N`` accepting an integer, with
            the number of nodes in the graph.
```

File: netket/graph/abstract_graph.py [hunk 6]
```diff
@@ -47,7 +47,8 @@ def nodes(self) -> Iterator[int]:
     @abc.abstractmethod
     def distances(self) -> List[List]:
         r"""List containing the distances between the nodes.
-        The fact that some node may not be reachable from another is represented by -1"""
+        The fact that some node may not be reachable from another is represented by -1
```
Comment (femtobit) [inline netket/graph/abstract_graph.py:50]:
```suggestion
        The distance between unconnected nodes (no path exists between them) is set to -1
```

**Review by femtobit — APPROVED** [review]
Looks good to me, except for some minor typos I've randomly noticed.

**Comment by PhilipVinc** [comment]
```
netket on  flake8 [⇣$] via netket_env via 🐍 system 
➜ flakehell lint netket

netket/graph/lattice.py
   556:  10 F821 undefined name 'SpaceGroupBuilder' [pyflakes]
  ) -> "SpaceGroupBuilder":
       ^

netket/logging/json_log.py
   130:  13 F841 local variable 'starting_json_content' is assigned to but never used [pyflakes]
  starting_json_content = json.load(open(output_prefix + ".log"))
  ^
```
